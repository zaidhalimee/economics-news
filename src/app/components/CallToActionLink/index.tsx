/** @jsx jsx */
/* @jsxFrag React.Fragment */
import React, { PropsWithChildren } from 'react';
import { jsx } from '@emotion/react';
import useClickTrackerHandler from '#app/hooks/useClickTrackerHandler';
import Text from '../Text';
import styles from './index.styles';
import { CallToActionLinkProps } from './types';

const CallToActionLink = ({
  href,
  className,
  children,
  eventTrackingData,
  size = 'pica',
  fontVariant = 'sansBold',
  download = false,
  ignoreLiteExtension = false,
  linkText = '',
  borderStyleOverride = false,
}: PropsWithChildren<CallToActionLinkProps>) => {
  const clickTrackerHandler = useClickTrackerHandler(eventTrackingData);

  return (
    <a
      href={href}
      className={className}
      css={styles.linkBackground}
      onClick={clickTrackerHandler}
      download={download}
      {...(ignoreLiteExtension && { 'data-ignore-lite': true })}
    >
      {borderStyleOverride ? (
        <>
          <Text size={size} fontVariant={fontVariant}>
            {linkText}
          </Text>
          {children}
        </>
      ) : (
        <div css={styles.linkTextWrapper}>
          <Text size={size} fontVariant={fontVariant} css={styles.linkText}>
            {children}
          </Text>
        </div>
      )}
    </a>
  );
};

export default CallToActionLink;
