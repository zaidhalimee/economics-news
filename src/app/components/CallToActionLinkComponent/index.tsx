/** @jsx jsx */
import { PropsWithChildren } from 'react';
import { jsx } from '@emotion/react';
import useClickTrackerHandler from '#app/hooks/useClickTrackerHandler';
import { EventTrackingMetadata } from '#app/models/types/eventTracking';
import Chevron from './Chevron';
import FlexWrapper from './FlexWrapper';

type CallToActionLinkProps = {
  to?: string;
  className?: string;
  eventTrackingData?: EventTrackingMetadata;
};

const CallToActionLink = ({
  to,
  children,
  eventTrackingData,
  ...htmlAttributes // handles LiteSiteCta
}: PropsWithChildren<CallToActionLinkProps>) => {
  const clickTrackerHandler = useClickTrackerHandler(eventTrackingData);
  const onClick = eventTrackingData ? clickTrackerHandler : () => null;

  return (
    <a href={to} onClick={onClick} {...htmlAttributes}>
      {children}
    </a>
  );
};

CallToActionLink.Chevron = Chevron;
CallToActionLink.FlexWrapper = FlexWrapper;

export default CallToActionLink;
