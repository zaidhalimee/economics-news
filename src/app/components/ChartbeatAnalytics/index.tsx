import React, { useContext } from 'react';
import useToggle from '../../hooks/useToggle';
import { RequestContext } from '../../contexts/RequestContext';
import { ServiceContext } from '../../contexts/ServiceContext';
import AmpChartbeatBeacon from './amp';
import CanonicalChartbeatBeacon from './canonical';
import { GetConfigProps, getConfig } from './utils';
import { ChartbeatProps } from './types';

const ChartbeatAnalytics = ({
  sectionName,
  categoryName,
  mediaPageType,
  title,
  taggings,
  contentType,
  producer,
  chapter,
}: ChartbeatProps) => {
  const { service, brandName, chartbeatDomain } = useContext(ServiceContext);
  const { env, isAmp, platform, pageType, previousPath, origin } =
    useContext(RequestContext);

  const { enabled } = useToggle('chartbeatAnalytics');

  if (!enabled) return null;

  const configDependencies: GetConfigProps = {
    isAmp,
    platform,
    pageType,
    brandName,
    chartbeatDomain,
    env,
    service,
    origin,
    previousPath,
    sectionName,
    categoryName,
    mediaPageType,
    title,
    taggings,
    contentType,
    producer,
    chapter,
  };

  const chartbeatConfig = getConfig(configDependencies);

  if (isAmp) return <AmpChartbeatBeacon chartbeatConfig={chartbeatConfig} />;

  return <CanonicalChartbeatBeacon chartbeatConfig={chartbeatConfig} />;
};

export default ChartbeatAnalytics;
