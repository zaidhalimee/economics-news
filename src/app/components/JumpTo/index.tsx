/* eslint-disable react-hooks/rules-of-hooks */
/** @jsx jsx */
import { useContext, useState, useEffect } from 'react';
import { jsx } from '@emotion/react';
import { ServiceContext } from '#contexts/ServiceContext';
import useViewTracker from '#app/hooks/useViewTracker';
import useClickTrackerHandler from '#app/hooks/useClickTrackerHandler';
import Text from '#app/components/Text';
import isLive from '#app/lib/utilities/isLive';
import { EventTrackingMetadata } from '#app/models/types/eventTracking';
import { OptimizelyContext } from '@optimizely/react-sdk';
import idSanitiser from '../../lib/utilities/idSanitiser';
import styles from './index.styles';

export interface JumpToProps {
  jumpToHeadings?: Array<{ heading: string }>;
}

const JumpTo = ({ jumpToHeadings }: JumpToProps) => {
  const { optimizely } = useContext(OptimizelyContext);

  const eventTrackingData: EventTrackingMetadata = {
    componentName: 'jumpto',
    optimizely,
    optimizelyEventName: 'jumpto',
  };

  // TODO: Remove for release
  if (isLive()) return null;

  const { translations } = useContext(ServiceContext);
  const [hash, setHash] = useState('');
  const { jumpTo = 'Jump to' } = translations?.articlePage || {};

  const viewRef = useViewTracker(eventTrackingData);
  const clickTrackerHandler = useClickTrackerHandler(eventTrackingData);

  useEffect(() => {
    setHash(window.location.hash);
  }, []);

  const linkClickHandler = (
    e: React.MouseEvent<HTMLElement, MouseEvent>,
    subheadingId: string,
  ) => {
    clickTrackerHandler(e);
    setHash(subheadingId);
  };

  const titleId = 'jump-to-heading';

  return (
    <nav
      ref={viewRef}
      role="navigation"
      aria-labelledby={titleId}
      data-testid="jump-to"
      css={styles.wrapper}
    >
      <Text
        as="strong"
        id={titleId}
        size="doublePica"
        fontVariant="sansBold"
        css={styles.title}
      >
        {jumpTo}
      </Text>
      <ol role="list" css={styles.list}>
        {jumpToHeadings?.map(({ heading }) => {
          const sanitisedId = idSanitiser(heading);
          const idWithHash = `#${sanitisedId}`;

          const isActiveId = decodeURIComponent(hash) === idWithHash;
          return (
            <li key={idWithHash} css={styles.listItem}>
              <a
                href={idWithHash}
                onClick={e => linkClickHandler(e, idWithHash)}
                css={styles.link}
                data-testid={`jump-to-link-${sanitisedId}`}
              >
                <span
                  css={[styles.linkText, isActiveId && styles.linkTextActive]}
                >
                  {heading}
                </span>
              </a>
            </li>
          );
        })}
      </ol>
    </nav>
  );
};

export default JumpTo;
