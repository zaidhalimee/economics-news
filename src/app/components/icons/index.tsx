import React from 'react';

export const Ellipsis = () => (
  <svg
    viewBox="0 0 32 32"
    focusable="false"
    aria-hidden="true"
    width="12"
    height="12"
  >
    <path d="M1.6 18.8h5.8v-5.6H1.6v5.6zm11.5 0h5.8v-5.6h-5.8v5.6zm11.5 0h5.8v-5.6h-5.8v5.6z" />
  </svg>
);
export const LeftChevron = ({ className }: { className?: string }) => (
  <svg
    viewBox="0 0 32 32"
    focusable="false"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M10.4 14.3L26.5 31h-6.4L5.5 16 20.1 1h6.4L10.4 17.7v-3.4z" />
  </svg>
);

export const RightChevron = ({ className }: { className?: string }) => (
  <svg
    viewBox="0 0 32 32"
    focusable="false"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M21.6 14.3L5.5 31h6.4l14.6-15L11.9 1H5.5l16.1 16.7v-3.4z" />
  </svg>
);

export const Clock = ({ className }: { className?: string }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 32 32"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M16 31c8.5 0 15-6.5 15-15S24.5 1 16 1 1 7.5 1 16s6.5 15 15 15zm0-2.7C9 28.3 3.7 23 3.7 16S9 3.7 16 3.7C23 3.7 28.3 9 28.3 16S23 28.3 16 28.3zm6.2-6.7 1-1.5-5.7-4.5-.6-8.6H15l-.7 10.5 7.9 4.1z" />
  </svg>
);

export const Book = ({ className }: { className?: string }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 32 32"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M8.6 1.6c-3.1 0-5 1.8-5 5v18.9c0 3.1 1.8 5 5 5h19.7V6.1h-2.2v21.7H8.2c-1.2 0-2.3-1-2.3-2.3s1-2.3 2.3-2.3h15.3V1.6H8.6z" />
  </svg>
);

export const Words = ({ className }: { className?: string }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 32 32"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M6 15.9h10v2H6zm4-4h12v2H10zm8 4h8v2h-8z" />
    <path d="M28 3.9H4c-1.1 0-2 .9-2 2v15c0 1.1.9 2 2 2h10.2v7l7.4-7H28c1.1 0 2-.9 2-2v-15c0-1.1-.9-2-2-2zm-6.4 17H20L16.2 25v-4.1H4v-15h24v15h-6.4z" />
  </svg>
);

export const Articles = ({ className }: { className?: string }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 32 32"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M5 9v20h22V9Zm20 18h-3a5.89 5.89 0 0 0-6-5 5.89 5.89 0 0 0-6 5H7V11h18ZM7 6h18v2H7zm2-3h14v2H9z" />
    <circle cx="15.998" cy="17.5" r="3.5" />
  </svg>
);

export const Calendar = ({ className }: { className?: string }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 32 32"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M6.2 30.5h19.5c5 0 5.3-2.5 5.3-5.3V9.3h-2.6v16c0 2-.6 2.6-2.6 2.6H6.3c-2 0-2.6-.6-2.6-2.6v-16H1v15.9c0 2.8.2 5.3 5.2 5.3zM10.8 15h2v-2h-2v2zM1 9.6h30V1.5h-5.3v3h-3v-3H9.2v3h-3v-3H1v8.1zM15 15h2v-2h-2v2zm4.1 0h2v-2h-2v2zm4.2 0h2v-2h-2v2zm-12.5 4.7h2v-2h-2v2zm4.2 0h2v-2h-2v2zm4.1 0h2v-2h-2v2zm4.2 0h2v-2h-2v2zm-16.6 0h2v-2h-2v2zm4.1 4.7h2v-2h-2v2zm4.2 0h2v-2h-2v2zm4.1 0h2v-2h-2v2zm-12.4 0h2v-2h-2v2z" />
  </svg>
);

export const Favourites = ({ className }: { className?: string }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 32 32"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M9.1 7.5h21.8V1.9H9.1v5.6zm0 22.6h21.8v-5.6H9.1v5.6zm7.2-11.3h14.6v-5.6H16.3v5.6zm-2.8-4.6H9.1V9.8H5.5v4.4H1.1v3.5h4.4v4.4h3.6v-4.4h4.4v-3.5z" />
  </svg>
);

export const Calculator = ({ className }: { className?: string }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 32 32"
    aria-hidden="true"
    width="12"
    height="12"
    className={className}
  >
    <path d="M22.8 32H9.2c-1.2 0-2.1-.3-2.6-.8s-.8-1.4-.8-2.6V3.4c0-1.2.3-2.1.8-2.6S8 0 9.2 0h13.6c1.3 0 2.1.3 2.6.8s.8 1.4.8 2.6v25.2c0 1.2-.3 2.1-.8 2.6s-1.3.8-2.6.8zm.8-29.4H8.4V12h15.2V2.6zM12.1 14.2H8.4v3.7h3.7v-3.7zm0 5.7H8.4v3.7h3.7v-3.7zm0 5.8H8.4v3.7h3.7v-3.7zm5.7-11.5h-3.7v3.7h3.7v-3.7zm0 5.7h-3.7v3.7h3.7v-3.7zm0 5.8h-3.7v3.7h3.7v-3.7zm5.8-11.5h-3.7v3.7h3.7v-3.7zm0 5.7h-3.7v9.5h3.7v-9.5z" />
  </svg>
);
