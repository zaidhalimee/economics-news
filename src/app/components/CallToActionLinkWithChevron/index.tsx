/** @jsx jsx */
import { PropsWithChildren, useContext } from 'react';
import { jsx, SerializedStyles, Theme } from '@emotion/react';
import { ServiceContext } from '#app/contexts/ServiceContext';
import { CallToActionLinkProps } from '../CallToActionLink/types';
import { LeftChevron, RightChevron } from '../icons';
import CallToActionLink from '../CallToActionLink';
import Text from '../Text';
import styles from './index.styles';

const CallToActionLinkWithChevron = ({
  href,
  className,
  children,
  eventTrackingData,
  download = false,
  chevronStyles,
}: PropsWithChildren<
  CallToActionLinkProps & { chevronStyles?: (_: Theme) => SerializedStyles }
>) => {
  const { dir } = useContext(ServiceContext);
  const isRtl = dir === 'rtl';

  return (
    <CallToActionLink
      href={href}
      className={className}
      eventTrackingData={eventTrackingData}
      download={download}
      css={styles.link}
    >
      <Text fontVariant="sansBold" css={styles.linkText}>
        {children}
      </Text>
      {isRtl ? (
        <LeftChevron css={chevronStyles} />
      ) : (
        <RightChevron css={chevronStyles} />
      )}
    </CallToActionLink>
  );
};

export default CallToActionLinkWithChevron;
