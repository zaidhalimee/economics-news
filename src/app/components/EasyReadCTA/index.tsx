/** @jsx jsx */
import { useContext } from 'react';
import { jsx } from '@emotion/react';
import { GridItemMedium } from '#app/legacy/components/Grid';
import { RequestContext } from '#app/contexts/RequestContext';
import useClickTrackerHandler from '#app/hooks/useClickTrackerHandler';
import useViewTracker from '#app/hooks/useViewTracker';
import Text from '../Text';
import { ServiceContext } from '../../contexts/ServiceContext';
import styles from './index.styles';
import { defaultTranslations } from './config';
import InlineLink from '../InlineLink';
import { FormatIcon } from '../icons';

type CtaLinkProps = {
  href: string;
  text: string;
  ignoreLiteExtension?: boolean;
  className?: string;
  selected?: boolean;
};

export const createHrefRelativeToPage = (currentPath: string, id?: string) => {
  const noExtenstionPath = currentPath.split('.')[0];
  const slugs = noExtenstionPath.split('/');
  const idToAppend = id ?? slugs[slugs.length - 1];
  slugs.splice(-1, 1, idToAppend);

  return slugs.join('/').replace(' ', '');
};

const eventTrackingData = {
  componentName: 'easyReadCta',
};

const CtaLink = ({ href, text, className, selected = false }: CtaLinkProps) => {
  const clickTrackerHandler = useClickTrackerHandler(eventTrackingData);

  return (
    <a
      href={href}
      className={className}
      css={styles.linkContainer}
      onClick={clickTrackerHandler}
      {...(selected && { 'aria-current': 'page' })}
    >
      <span css={styles.linkTextContainer}>
        <Text
          size="brevier"
          css={[
            styles.linkText,
            selected ? styles.selected : styles.notSelected,
          ]}
        >
          {text}
        </Text>
      </span>
    </a>
  );
};

type Props = {
  easyReadAssetId?: string;
  originalAssetId?: string;
};

const EasyReadCTA = ({ easyReadAssetId, originalAssetId }: Props) => {
  const viewRef = useViewTracker(eventTrackingData);
  const { pathname } = useContext(RequestContext);
  const { translations } = useContext(ServiceContext);

  if (easyReadAssetId == null && originalAssetId == null) {
    return null;
  }

  const easyHref = createHrefRelativeToPage(pathname, easyReadAssetId);
  const originalHref = createHrefRelativeToPage(pathname, originalAssetId);

  const isEasyActive = easyReadAssetId == null;

  const { easyReadSite = defaultTranslations } = translations;
  const {
    toStandardSite,
    toEasySite,
    format,
    aIDisclaimer,
    learnMore,
    learnMoreLink,
  } = easyReadSite;

  const id = 'Format';

  return (
    <GridItemMedium>
      <section
        role="region"
        data-e2e="easy-read-cta"
        aria-labelledby={id}
        ref={viewRef}
      >
        <FormatIcon css={styles.icon} />
        <Text as="strong" id={id} hidden>
          {format}
        </Text>
        <CtaLink
          href={originalHref}
          text={toStandardSite}
          {...(!isEasyActive && { selected: true })}
        />
        <CtaLink
          href={easyHref}
          text={toEasySite}
          {...(isEasyActive && { selected: true })}
        />
      </section>
      {isEasyActive && (
        <Text as="small" size="brevier" css={styles.disclaimer}>
          {aIDisclaimer} <InlineLink text={learnMore} to={learnMoreLink} />.
        </Text>
      )}
    </GridItemMedium>
  );
};

export default EasyReadCTA;
