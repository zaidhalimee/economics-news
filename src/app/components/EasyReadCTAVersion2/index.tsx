/** @jsx jsx */
import { useContext } from 'react';
import { jsx } from '@emotion/react';
import Text from '../Text';
import { ServiceContext } from '../../contexts/ServiceContext';
import styles from './index.styles';
import { defaultTranslations } from './config';
import InlineLink from '../InlineLink';
import { FormatIcon } from '../icons';

type CtaLinkProps = {
  href: string;
  text: string;
  ignoreLiteExtension?: boolean;
  className?: string;
  selected?: boolean;
};

const CtaLink = ({ href, text, className, selected = false }: CtaLinkProps) => {
  return (
    <a
      href={href}
      className={className}
      css={styles.link}
      {...(selected && { 'aria-current': 'page' })}
    >
      <Text size="brevier" css={[styles.linkText, selected && styles.selected]}>
        {text}
      </Text>
    </a>
  );
};

type Props = {
  easyVersionLink?: string;
  fullVersionLink?: string;
};

const EasyReadCTA = ({ easyVersionLink, fullVersionLink }: Props) => {
  const isEasyVersion = fullVersionLink != null;

  const { translations } = useContext(ServiceContext);

  const { easyReadSite = defaultTranslations } = translations;
  const {
    toStandardSite,
    toEasySite,
    format,
    aIDisclaimer,
    learnMore,
    learnMoreLink,
  } = easyReadSite;

  const id = 'Format';
  const href = isEasyVersion ? fullVersionLink : easyVersionLink;
  if (href == null) return null;

  return (
    <div css={styles.bottomSpacing}>
      <section role="region" data-e2e="easy-read-cta" aria-labelledby={id}>
        <FormatIcon css={styles.icon} />
        <Text as="strong" id={id} hidden>
          {format}
        </Text>
        <CtaLink href={href} text={toStandardSite} />
        <CtaLink href={href} text={toEasySite} selected />
      </section>
      {isEasyVersion && (
        <Text as="small" size="brevier" css={styles.disclaimer}>
          {aIDisclaimer} <InlineLink text={learnMore} to={learnMoreLink} />.
        </Text>
      )}
    </div>
  );
};

export default EasyReadCTA;
