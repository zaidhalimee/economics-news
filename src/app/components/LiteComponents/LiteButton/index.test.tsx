/* eslint-disable no-console */
import React from 'react';
import { Helmet } from 'react-helmet';
import { render } from '#app/components/react-testing-library-with-providers';
import LiteButton from '.';

jest.mock('react', () => {
  const ActualReact = jest.requireActual('react');
  return {
    ...ActualReact,
    useId: jest.fn().mockReturnValue(':r0:'),
  };
});

describe('LiteButton', () => {
  it('should render a script tag with the passed in script string', () => {
    render(
      <LiteButton script="console.log(`Hello, world! I'm a string function`)">
        Click me
      </LiteButton>,
    );

    const helmet = Helmet.peek();
    expect(helmet.scriptTags).toHaveLength(1);

    const script = helmet.scriptTags[0].innerHTML;
    expect(script).toMatchSnapshot();
  });

  it('should render a script tag with the passed in script function', () => {
    render(
      <LiteButton
        script={() =>
          console.log(
            `Hello, world! I'm an arrow function transpiled to a traditional function`,
          )
        }
      >
        Click me
      </LiteButton>,
    );

    const helmet = Helmet.peek();
    expect(helmet.scriptTags).toHaveLength(1);

    const script = helmet.scriptTags[0].innerHTML;

    expect(script).toMatchSnapshot();
  });
});
