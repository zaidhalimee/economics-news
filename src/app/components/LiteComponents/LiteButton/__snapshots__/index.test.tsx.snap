// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiteButton should render a script tag with the passed in script function 1`] = `
"
    window.addEventListener("load", function(event) {
      var button = document.getElementById('lite-button-:r0:');
      button.addEventListener('click', function script() {
        return console.log("Hello, world! I'm an arrow function transpiled to a traditional function");
      });
    });
  "
`;

exports[`LiteButton should render a script tag with the passed in script string 1`] = `
"
      window.addEventListener("load", function(event) {
        var button = document.getElementById('lite-button-:r0:');
        button.addEventListener('click', function() {console.log(\`Hello, world! I'm a string function\`)});
      });
    "
`;
